apis/stacks/v1alpha1/stackdefinition.go

```diff
@@ -0,0 +1,110 @@
+	Permissions     PermissionsSpec         `json:"permissions,omitempty"`
+	Behavior        StackDefinitionBehavior `json:"behavior,omitempty"`
+}
+
+// StackDefinitionSource is the stack image which this stack configuration is from.
+// In the future, other source types may be supported, such as a URL.
+type StackDefinitionSource struct {
+	// a container image id
+	Image string `json:"image,omitempty"`
+}
+
+// StackDefinitionBehavior specifies the behavior for the stack, if the stack has behaviors instead of a controller
+type StackDefinitionBehavior struct {
+	CRD    StackDefinitionBehaviorCRD       `json:"crd,omitempty"`
+	Engine StackResourceEngineConfiguration `json:"engine,omitempty"`
+	// Theoretically, source and engine could be specified at a per-hook level
+	// as well.
+	Source    StackDefinitionSource              `json:"source,omitempty"`
+	Reconcile []StackDefinitionHookConfiguration `json:"reconcile,omitempty"`
+}
+
+// StackDefinitionBehaviorCRD represents the CRD which the stack's behavior controller will watch. When CRs of this
+// CRD kind appear and are modified in the cluster, the behavior will execute.
+type StackDefinitionBehaviorCRD struct {
+	APIVersion string `json:"apiVersion"`
+	Kind       string `json:"kind"`
+}
+
+// StackResourceEngineConfiguration represents a configuration for a resource engine, such as helm2 or kustomize.
+type StackResourceEngineConfiguration struct {
+	Type     string                  `json:"type"`
+	Overlays []ResourceEngineOverlay `json:"overlays,omitempty"`
+}
+
+// ResourceEngineOverlay configures the stack behavior controller to transform the input CR into some output objects
+// for the underlying resource engine. This is expected to be interpreted by the engine-specific logic in the controller.
+type ResourceEngineOverlay struct {
+	APIVersion string                         `json:"apiVersion"`
+	Kind       string                         `json:"kind"`
+	Name       string                         `json:"name"`
+	Bindings   []ResourceEngineOverlayBinding `json:"bindings"`
+}
+
+// ResourceEngineOverlayBinding describes a field binding of an overlay transformation. It connects a field in the source
+// object to a field in the destination object.
+type ResourceEngineOverlayBinding struct {
+	From string `json:"from"`
+	To   string `json:"to"`
+}
+
+// StackDefinitionHookConfiguration is the configuration for an individual hook which will be
+// executed in response to an event.
+type StackDefinitionHookConfiguration struct {
+	Path string `json:"path"`
+}
+
+// StackDefinitionStatus defines the observed state of StackDefinition
+type StackDefinitionStatus struct {
+	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+}
+
+// +kubebuilder:object:root=true
+
+// StackDefinition is the Schema for the StackDefinitions API
+type StackDefinition struct {
```

Comment:
This comment should be irrelevant as I understand https://github.com/crossplaneio/crossplane/issues/1209#issuecomment-579568139 after the review.

Reply:
Yeah, it's because of the thinking in the comment you linked that this is being done.

There will still be a question for the future of whether we want users to be interacting with the same type as what the system is using underneath. For example, if we are generating controller deployment blocks, maybe we don't want to expose the user to the complexity of that unnecessarily. But I believe we can revisit this after a little while when we want to improve the authoring experience and support various authoring modes.